src/Compilers/CSharp/Portable/Symbols/TypeSymbolWithAnnotations.cs

```diff
@@ -639,7 +606,12 @@ public bool Equals(TypeSymbolWithAnnotations other, TypeCompareKind comparison)
                 return true;
             }
 
-            if (other.IsNull || !TypeSymbolEquals(other, comparison))
+            if (!HasType)
+            {
+                if (other.HasType || NullableAnnotation != other.NullableAnnotation)
```

Comment:
I think we can just return `!other.HasType && NullableAnnotation == other.NullableAnnotation`, right?